src/realm/replication.cpp

```diff
@@ -62,36 +64,323 @@ Replication::version_type Replication::prepare_commit(version_type orig_version)
+}
+
+void Replication::nullify_link(const Table* t, ColKey col_key, ObjKey key)
+{
+    select_table(t); // Throws
+    select_obj(key);
+    m_encoder.modify_object(col_key, key); // Throws
+    if (auto logger = get_logger()) {
+        logger->log(util::Logger::Level::trace, "   Nullify '%1'", t->get_column_name(col_key));
+    }
+}
+
+void Replication::add_int(const Table* t, ColKey col_key, ObjKey key, int_fast64_t value)
+{
+    do_set(t, col_key, key); // Throws
+    if (auto logger = get_logger()) {
+        logger->log(util::Logger::Level::trace, "   Adding %1 to '%2'", value, t->get_column_name(col_key));
+    }
+}
+
+
+Path Replication::get_prop_name(Path&& path) const
+{
+    Path ret(std::move(path));
+    auto col_key = ret[0].get_col_key();
+    auto prop_name = m_selected_table->get_column_name(col_key);
+    ret[0] = PathElement(prop_name);
+    return ret;
+}
+
+void Replication::log_collection_operation(const char* operation, const CollectionBase& collection, Mixed value,
+                                           Mixed index) const
+{
+    auto logger = get_logger();
+    auto path = collection.get_short_path();
+    auto col_key = path[0].get_col_key();
+    auto prop_name = m_selected_table->get_column_name(col_key);
+    path[0] = PathElement(prop_name);
+    std::string position;
+    if (!index.is_null()) {
+        position = util::format(" at position %1", index);
+    }
+    if (Table::is_link_type(col_key.get_type()) && value.is_type(type_Link)) {
+        auto target_table = m_selected_table->get_opposite_table(col_key);
+        if (target_table->is_embedded()) {
+            logger->log(util::Logger::Level::trace, "   %1 embedded object '%2' in %3%4 ", operation,
+                        target_table->get_class_name(), path, position);
+        }
+        else if (target_table->get_primary_key_column()) {
+            auto link = value.get<ObjKey>();
+            auto pk = target_table->get_primary_key(link);
+            logger->log(util::Logger::Level::trace, "   %1 object '%2' with primary key %3 in %4%5", operation,
+                        target_table->get_class_name(), pk, path, position);
+        }
+        else {
+            auto link = value.get<ObjKey>();
+            logger->log(util::Logger::Level::trace, "   %1 object '%2'[%3] in %4%5", operation,
+                        target_table->get_class_name(), link, path, position);
+        }
+    }
+    else {
+        logger->log(util::Logger::Level::trace, "   %1 %2 in %3%4", operation, value, path, position);
+    }
+}
+void Replication::list_insert(const CollectionBase& list, size_t list_ndx, Mixed value, size_t)
+{
+    select_collection(list);                                     // Throws
+    m_encoder.collection_insert(list.translate_index(list_ndx)); // Throws
+    if (auto logger = get_logger()) {
+        if (logger->would_log(util::Logger::Level::trace)) {
+            log_collection_operation("Insert", list, value, int64_t(list_ndx));
```

Comment:
Should we prepend the collection type to the operation type so we can easily distinguish them? (eg, `ListInsert`)

Reply:
I hope it will not be needed.